Modifies WPAnimatedBox to prevent endless animation

stevebaranski opened the pull request:
## Description
Fixes #10047, which notes high CPU utilization in the _Reader_ related to `WPAnimatedBox`.

### Root cause

1. Recently this custom view was modified to loop continuously.
1. The custom view was proliferated with the addition of "empty state" views employing `NoResultsViewController` (including _Reader_).

### Resolution
The following changes were made to address the problem:

- Simplified `-[WPAnimatedBox animate]` method.
- Exposed additional `-[WPAnimatedBox suspendAnimation]` method to end the animation loop.
- Removed `WPNoResultsView+AnimatedBox`, which was no longer in use.
- Modified `NoResultsViewController` and its _explicit_ callers to manage the animation lifecycle. It's worth noting that additional _implicit_ callers benefit from the change to `class NoResultsViewController.loadingAccessoryView()`.

### Future improvements
While the solution proposed appears to resolve the issue, it does require that future users of `WPAnimatedBox` & `NoResultsViewController` follow a convention of sorts. 

In the future, we might consider additional safeguards. For example:
- We might designate a maximum animation duration or loop count for `WPAnimatedBox`
- We might improve the handling of `accessoryView`;  we could clarify the type (currently `UIView!` or even suspend animations if a `WPAnimatedBox` is set where animations are currently in-flight.

## To test

- Checkout branch and verify that existing tests pass.
- Because this is related to CPU utilization, this will be easier to test in an Xcode simulator. To do so: run the app in a variety of scenarios (e.g., signed in, out, different accounts); exercise a variety of _No Results_ views, including the Reader; verify that CPU utilization in _Debug navigator_ and/or Instruments continues to behave.
- It's possible that I might have missed a _No Results_ code path; please let me know if you see anything!

cc: @loremattei for v10.8 visibility

stevebaranski commented:
Thanks for the input @ScoutHarris!

Also, I elected to refer the removal of `WPNoResultsView+AnimatedBox` from this PR, to reduce surface area re: this change into a release branch post-freeze. You should find no ill-effects.

ScoutHarris commented on WordPress/Classes/ViewRelated/Blog/NoResultsViewController.swift:
```diff
@@ -263,4 +270,21 @@ private extension NoResultsViewController {
+    // MARK: - `WPAnimatedBox` resource management
+
+    private func startAnimatingIfNeeded() {
+        guard let animatedBox = accessorySubview as? WPAnimatedBox else { return }
```
Code style nitpick: the return should be on a separate line, i.e.
```
... else {
return 
}
```

nheagy commented on WordPress/Classes/ViewRelated/Blog/NoResultsViewController.swift:
```diff
@@ -263,4 +270,21 @@ private extension NoResultsViewController {
+    // MARK: - `WPAnimatedBox` resource management
+
+    private func startAnimatingIfNeeded() {
+        guard let animatedBox = accessorySubview as? WPAnimatedBox else { return }
```
The `return` should [be on its own line](https://github.com/wordpress-mobile/swift-style-guide#…)

stevebaranski replied:
I'm happy to make the change. Can this be properly enforced with SwiftLint?

nheagy commented on WordPress/Classes/ViewRelated/Blog/NoResultsViewController.swift:
```diff
@@ -263,4 +270,21 @@ private extension NoResultsViewController {
+    }
+
+    private func stopAnimatingIfNeeded() {
+        guard
```
The formatting of this `guard` statement looks weird. Have we done this anywhere else?

ScoutHarris replied:
Typically we have the first `let` on the same line as the `guard`.

stevebaranski replied:
I have used this formatting in a few places, but I'm happy to make the change.

ScoutHarris commented on WordPress/Classes/ViewRelated/Blog/NoResultsViewController.swift:
```diff
@@ -263,4 +270,21 @@ private extension NoResultsViewController {
+    private func stopAnimatingIfNeeded() {
+        guard
+            let accessorySubview = accessorySubview,
+            let animatedBox = accessorySubview as? WPAnimatedBox
```
Can these two `let`s be consolidated into one line, like above in `startAnimatedIfNeeded`?

`guard let animatedBox = accessorySubview as? WPAnimatedBox`

stevebaranski replied:
Good call! It also addresses some of the formatting feedback, albeit indirectly. I'm happy to make this change.

ScoutHarris commented on WordPress/Classes/ViewRelated/Blog/NoResultsViewController.swift:
```diff
@@ -263,4 +270,21 @@ private extension NoResultsViewController {
+            let accessorySubview = accessorySubview,
+            let animatedBox = accessorySubview as? WPAnimatedBox
+        else
+        {
```
Code style nitpick: This brace should be on the same line as the else.

stevebaranski commented:
@nheagy & @ScoutHarris : thank you for the prompt review, and for your input regarding formatting. I have pushed an additional commit that addresses the concerns you noted. When you have a moment, would you be able to give this another look?

ScoutHarris commented:
Thank you @stevebaranski ! :shipit:
